Sync samples with Lims on proposal selection

matyson opened the pull request:
Closes #1981 . I chose to handle the sync directly in SelectProposal dialog since both `selectProposal`and `getLimsSamples` actions were already implemented.

marcus-oscarsson commented:
Yes, I guess that makes sense. The only sort of problem I see is that the user never chooses which LIMS to synchronize with. I guess is not really an issue, (@antolinos what do you think ?). In the case of ISPyB the user can simply make another specific sync

antolinos commented:
I believe it is a good idea and will work for many of you, but I wonder whether it should be configurable. In the ESRF case, I see a couple of drawbacks:
- People belonging to a proposal use multiple LIMS, often simultaneously. Some pucks are registered in ISPyB and others in DRAC, and users then end up synchronizing twice. This happens especially in BAGs, where some people are moving to DRAC while others are still using ISPyB. One could argue that once we phase out ISPyB this will no longer be a problem, which I agree with. However, at some point everybody will abandon ISPyB (or whatever lims is currently used), and this dual mechanism will be very useful to keep.
- Synchronization is fast… but not always. Now we can associate search models and reference files with crystals, and this data needs to be downloaded during synchronization. It may be slow (a few seconds). I think this is acceptable during sample synchronization but not when selecting the proposals.
- You might want to collect samples manually or perform XRF or energy scans; in these cases, you don’t need to synchronize pucks. So why force it?

For the future, these are ~requirements~  wishes hat I think will need to be implemented in a near future and having a specific button for synch might help:
- You might want to synchronize only certain samples. Example: synchronize only uncollected samples 
- You might want to put in the SC pucks from multiple proposals and collect them all together

These are just some ideas from an external point of view and are very biased by the way I see things

matyson commented:
I could add a configuration to enable autosync on proposal selection to keep the current behaviour as is, but i think even this won't attend the use cases above for multiple synching. In particular, we also have interest in multiple proposal selection, but i think would need to understand better how the synching between proposals/lims should be performed because even now the current "Sync button" in samples page only reproduces the same function call as listed in this PR. 

So, should we have more than one `limsName` available it should be selectable somewhere in the UI.

antolinos commented:
Hi again,

We currently have the option to choose the LIMS in the syncing button:
<img width="638" height="312" alt="Screenshot from 2025-12-10 15-50-51" src="https://github.com/user-attachments/assets/185b24a8-4320-432e-9cbc-256038a6ce45" />

Do you refer to this?

marcus-oscarsson commented:
It would be nice if it was configurable to sync on session selection for the time being. I think its something we would use but our current use cases are quite complicated and it would be nice to be able to opt out if needed. 

It would be nice if it could be something like: `None`, `ispyb`, `drac` 

Its true that this operation may take some time, so it would be really useful with a busy/loading screen for the proposal selection and sync process.

matyson commented:
> Hi again,
> 
> We currently have the option to choose the LIMS in the syncing button: <img alt="Screenshot from 2025-12-10 15-50-51" width="638" height="312" src="https://private-user-images.githubusercontent.com/2901169/524877937-185b24a8-4320-432e-9cbc-256038a6ce45.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjUzNzg3OTMsIm5iZiI6MTc2NTM3ODQ5MywicGF0aCI6Ii8yOTAxMTY5LzUyNDg3NzkzNy0xODViMjRhOC00MzIwLTQzMmUtOWNiYy0yNTYwMzhhNmNlNDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTIxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEyMTBUMTQ1NDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjYzYTM3MGRhMWY3YjI2OTliNzFlYTI0NGFmM2NjZjY1MWFjMWM0ZThjZmE5YTkzNDUyYzMwZjE1OWJiNDhhMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.MGRP9-qpEqPvaNrR10grCjhNM83x7pDkMNHTsmW-NRs">
> 
> Do you refer to this?

Ohh, you're right! I just misread it, my bad. So multiple lims samples as a config could be something like:

```yaml
autosync_lims:
  - ispyb
  - drac
```

Not sure about the proper naming conventions for these kind of configs. Also, for multiple lims are we merging the sample list? I only glanced at this [ISPyBAbstractLims](https://github.com/mxcube/mxcubecore/blob/develop/mxcubecore/HardwareObjects/abstract/ISPyBAbstractLims.py#L182) and since it just gets from one proposal at a time i'm assuming the facility specific ones merge the sample list..

marcus-oscarsson commented:
That would work, at the moment we only sync with one at the time but we can keep the option to have two if ever needed

marcus-oscarsson commented:
Just a question, does this really solve #1981 completely. The users can still retrieve the samples from the SC on the sample list view, see screen shot from @antolinos above. In that case it wont matter if the lims samples are retrieved at proposal selection, right ?

Would it not then make more sense to either remove or change the behavior of `Get samples from sc`. In your case you actually only need one option in the sample list view `Get samples` right ?

Just some thoughts after have talked to @antolinos

jbflo commented:
> Just a question, does this really solve #1981 completely. The users can still retrieve the samples from the SC on the sample list view, see screen shot from @antolinos above. In that case it wont matter if the lims samples are retrieved at proposal selection, right ?
> 
> Just some thoughts after have talked to @antolinos

@marcus-oscarsson
There is a config flag that hides the “Get samples from SC” option in the UI: USE_GET_SAMPLES_FROM_SC.

When it is set to False, users no longer have the option to retrieve raw SC samples, only the synchronized ones.

You can maybe use this boolean to control whether mxcube auto-syncs with a default LIMS on proposal selection or not.

> 
> Would it not then make more sense to either remove or change the behavior of `Get samples from sc`. In your case you actually only need one option in the sample list view `Get samples` right ?
> 
I think the current behavior of “Get samples from SC” is fine as it is.
We just need to better understand its use case and clearly define where/how and when it should be called.

Hope that helps!

jbflo commented on ui/src/components/LoginForm/SelectProposal.jsx:
```diff
@@ -72,6 +78,11 @@ function SelectProposal() {
+    dispatch(selectProposal(sessionId));
+    dispatch(getLimsSamples(limsName[0].name));
+  }
+
```
@matyson I think it’s worth checking the backend side for this.
With this approach, the backend does something and then you revert it in the UI. It’s probably cleaner to fix the behavior directly in the backend instead of compensating in the frontend.

In [mxcubeweb/core/components/lims.py around L411:](https://github.com/mxcube/mxcubeweb/blob/dc5753586b977360baca37b77c68fc5a35e0f28c/mxcubeweb/core/components/lims.py#L411)

`self.get_sample_list_from_sc()  # changing this one line might solve your issue
`

You could try replacing that with:

`self.synch_with_lims(limsname)
`

matyson replied:
So, what you're suggesting is to use this `USE_GET_SAMPLES_FROM_SC` directly in the `lims.py` to choose whether to sync from sc or the lims, right?

marcus-oscarsson replied:
I don't think its a good idea to mix these two functionalities, its two separate features. I think we need to take a step back and look at the use cases that you are trying to solve. To me it seems like you would like to prevent users from seeing/using samples that do not belong to them but are in the sample changer (for understandable reasons)

We should perhaps meet quickly to discuss the details a bit more closer ?

matyson replied:
> To me it seems like you would like to prevent users from seeing/using samples that do not belong to them but are in the sample changer 

Yes, that's my goal, but it seems to me that showing all the samples is a very used feature and therefore difficult to find a good solution for all. 

Absolutely, we can meet. What's your availability?

marcus-oscarsson replied:
I sent an email, this afternoon after 14h or tomorrow between 9h and 11h or after 13h

matyson committed:
add config to enable autosync LIMS samples on proposal selection

- update configmodels
- update demo files
- add missing docs
- add state to general reducer

matyson commented:
So, as we've discussed i added the configuration `AUTOSYNC_LIMS` as a boolean to app general configuration in the`server.yaml` file's `mxcube` section (not sure if it makes sense there though, i can change it if you see fit). 

Also, i took the initiative to add the `USE_GET_SAMPLES_FROM_SC` missing templates and docs as they seemed related enough.

matyson committed:
fetch LIMS on proposal selection (if enabled)
